R/causal_impact.R

```diff
@@ -0,0 +1,186 @@
+#'                      "model_stats" - model fit summary from broom::glance() on the bsts model.
+#'                      "model_coef" - model coefficients from broom::tidy() on the bsts model.
+#'                      "model" - model data frame with the bsts model. (Not use in Exploratory UI for now.)
+#' @param alpha - Tail-area probability of posterior interval (a concept that is similar to confidence interval.)
+#' @param niter - Number of MCMC Samples.
+#' @param standardize.data - Whether to standardize data.
+#' @param prior.level.sd - Prior Standard Deviation of Random Walk
+#' @param nseasons - Period of Seasonal Trend. e.g. 7, for weekly trend.
+#' @param season.duration - Used with nseasons. How many unit time one season consists of. e.g. 24, when unit time is hour.
+#' @param dynamic.regression - Whether to include time-varying regression coefficients.
+#' @param ... - extra values to be passed to CausalImpact::CausalImpact.
+do_causal_impact_ <- function(df, time_colname, formula, intervention_time = NULL, output_type = "series",
+                              niter = NULL, standardize.data = NULL, prior.level.sd = NULL, nseasons = NULL, season.duration = NULL, dynamic.regression = NULL, ...) {
+  y_colname <- as.character(lazyeval::f_lhs(formula))
+  predictor_column_names <- all.vars(lazyeval::f_rhs(formula))
+  all_column_names <- all.vars(formula)
+  grouped_col <- grouped_by(df)
+
+  # column name validation
+  if(!time_colname %in% colnames(df)){
+    stop(paste0(time_colname, " is not in column names"))
+  }
+
+  if(time_colname %in% grouped_col){
+    stop(paste0(time_colname, " is grouped. Please ungroup it."))
+  }
+
+  if (!class(df[[time_colname]]) %in% c("Date", "POSIXct")) {
+    stop(paste0(time_colname, " must be Date or POSIXct."))
+  }
+
+  if (!is.null(intervention_time) && class(intervention_time) != "character" && class(df[[time_colname]]) != class(intervention_time)) {
+    stop(paste0("intervention_time must be character or the same class as ", time_colname, "."))
+  }
+
+  # remove rows with NA in predictors. CausalImpact does not allow NA in predictors (covariates).
+  for(var in predictor_column_names) {
+    df <- df[!is.na(df[[var]]), ]
+  }
+  # remove NA data
+  df <- df[!is.na(df[[time_colname]]), ]
+
+  do_causal_impact_each <- function(df) {
+    # select only columns that appear in the formula.
+    # following fails with column names with spaces most likely because of dplyr bug.
+    # input_df <- dplyr::select_(df, .dots = c(time_colname, all_column_names))
+    # using base R style to avoid it.
+    # since this base R style "select" seems to mess up grouping info if done outside of do_causal_impact_each,
+    # it has to be done inside do_causal_impact_each.
+    input_df <- df[, c(time_colname, all_column_names)]
+    # rename the y column to fixed name y so that it is easier to handle in the next step.
+    input_df <- dplyr::rename_(input_df, y = y_colname)
```

Comment:
It's possible but I don't know the way to do  `dplyr::select(input_df, y, dplyr::everything())` without a fixed column name. Let me know if there is a better way.
For now, I renamed the name y to .y_value, which would have smaller possibility of conflict.